### ISSUE TITLE
Refactor cli arg parsing

### ISSUE THREAD

**Issue by deontologician** [rethinkdb/horizon#722]
Right now there's lots of duplicated copy/pasted code, and it's hard to follow. The ideal situation would be specifying in one place for each option:
### Option specification
1. The argument name
2. All argparse options
3. the default value
4. A namespace if necessary (`--rdb-*`, or `--auth-*`)
5. a function for any special post-processing
   - `bind=all` turns into `bind: [ '0.0.0.0' ]`
   - `connect` needs to be parsed afterward
   - if this post-processor fails, processing should stop and the exception should be shown in red
6. Whether the option is considered a secret (and should come from `secrets.toml`) (default:no)
7. Whether the option should be part of --dev (default: no)
   - this means every command can potentially have a `--dev` option, but they don't need to use it
8. Whether the option should be specifiable in the config (default: yes)
9. Whether the option should be specifiable in the command line (default: yes)
10. Whether the option should be specifiable as an env variable
11. An optional validation group for this option

Additionally, a validation function will need to be able to be associated with the validation group so that commands that potentially conflict (like --rdb-port and --connect) can be validated together once everything is merged into a final options object. The function would receive all values of commands in that group, even ones that defaulted.
### Features:
1. it should handle our `yes|no` boolean stuff transparently. If you specify an option `type: 'bool'` it will accept yes/no, there shouldn't be a need to do any more special checks.
2. it should handle env/config/secrets/option merging completely transparently, that can get out of the individual commands
3. it should automatically add any dev options to the `--dev` option help text and handle the logic for adding dev options
4. It should be easy to pull out reuseable options like rethinkdb options and auth options and just extend the per-command options very simply. This will also help plugin writers.
### misc cleanups
- `_implicit_start_rethinkdb` and `_dev_flag_used` express the same information and should be merged.
